packages/utils/src/useControlled.ts

```diff
@@ -3,6 +3,20 @@
 /* eslint-disable react-hooks/rules-of-hooks, react-hooks/exhaustive-deps */
 import * as React from 'react';
 
+/**
+ * Safely stringifies a value for use in useEffect dependencies.
+ * Falls back to reference comparison for circular structures.
+ */
+function safeStringifyForDeps(value: unknown): string | unknown {
+  try {
+    return JSON.stringify(value);
+  } catch {
+    // If JSON.stringify fails (e.g., circular structure),
+    // return the original value for reference comparison
+    return value;
+  }
+}
```

Comment:
If comparing by reference is correct, then why are we stringifying in the first place? Stringify has a possibly huge overhead compared to just a comparison by reference.

Reply:
This is an error - removed